### ISSUE TITLE
Training base model and X-Paste, performances

### ISSUE THREAD

**Issue by DavideA** [yoctta/XPaste#2]
Hi, 
and thank you for sharing your code. 

I am trying to reproduce results in the paper and I'm having issues with the following two configurations:

**Base Model**

I am submitting the training script with the config file `configs/Base-C2_L_R5021k_640b64_4x.yaml` to train the base CenterNetv2 (ResNet50 backbone). However, the results I get are slightly worse than expected, is this the right way to train it?

| Model | source| $AP^{box}$ | $AP^{mask}$ | $AP_r^{box}$ | $AP_r^{mask}$ |
|:--------:|:----------:|:----------:|:-----------:|:------------:|:-------------:|
| baseline  | checkpoint |    34.5    |     30.8    |     24.0     |      21.6    |
| baseline  | retraining |    34.0   |    30.0  |   21.7  |      18.4  |

**X-Paste model**
Same happens when I train the X-Paste model using the `configs/Xpaste_R50.yaml` config. Here differences are more evident.

| Model| source| $AP^{box}$ | $AP^{mask}$ | $AP_r^{box}$ | $AP_r^{mask}$ |
|:--------:|:----------:|:----------:|:-----------:|:------------:|:-------------:|
| XPaste | checkpoint |    37.4 |     33.2 |     33.9 |      29.7 |
| XPaste | retraining |    34.5   |    30.4  |   24.2   |      21.8 |

What am I missing?

Thanks for help,
D

**Comment by yoctta** [comment]
Hello, can you provide some details of your experiment? Training batch size, iteration and quantity of generated / retrieved samples.

**Comment by Z-MU-Z** [comment]
Here is my result using configs/Base-C2_L_R5021k_640b64_4x.yaml

|   AP   |  AP50  |  AP75  |  APs   |  APm   |  APl   |  APr   |  APc   |  APf   |
|:------:|:------:|:------:|:------:|:------:|:------:|:------:|:------:|:------:|
| 34.229 | 47.118 | 36.028 | 25.603 | 43.366 | 49.878 | 23.166 | 33.306 | 40.123 |
|   AP   |  AP50  |  AP75  |  APs   |  APm   |  APl   |  APr   |  APc   |  APf   |
|:------:|:------:|:------:|:------:|:------:|:------:|:------:|:------:|:------:|
| 30.404 | 44.566 | 32.534 | 21.529 | 40.919 | 46.623 | 20.895 | 29.923 | 35.119 |

**Comment by Z-MU-Z** [comment]
@yoctta 
I am not very clear on how to retrieve the images, which means I am unable to reproduce the results in the paper. Could you provide the corresponding data and code? I would be very grateful.

**Comment by yoctta** [comment]
> @yoctta I am not very clear on how to retrieve the images, which means I am unable to reproduce the results in the paper. Could you provide the corresponding data and code? I would be very grateful.

We retrieve the Images from search engine. I cannot provide the data or script now since I had left Microsoft. If you want to require the data, you may contact other authors.

**Comment by DavideA** [comment]
> Hello, can you provide some details of your experiment? Training batch size, iteration and quantity of generated / retrieved samples.

@yoctta Thank you for your answer. I am following the configuration files in `configs/Base-C2_L_R5021k_640b64_4x.yaml` and `configs/Xpaste_R50.yaml`. As such:

- For the base CenterNetv2 with ResNet 50 backbone, the batch size is 64 and iterations are 90k.
- For the XPaste version, the batch size is 64 and iterations are 180k (with EMA).

As for the pool, I am simply replicating the one proposed in your README. Basically, I use only generated samples, no retrieved instances, and I generate 100 per class. This gives 1203*100=120333 generated instances, that become 107242 after the CLIP filtering.

Aren't the pipeline in the README and the configuration files supposed to replicate the results of the released checkpoints?

**Comment by yoctta** [comment]
we use configs/Base-C2_L_R5021k_640b64_4x.yaml with 90k iters to train the base model. Which doesn't gain from more training steps. And use the model to initialize Xpaste training.

**Comment by leaf1170124460** [comment]
Hi, @yoctta , I use below config to train a baseline swinL， get much better result than the paper‘s.


_BASE_: "./Base-C2_L_R5021k_640b64_4x.yaml"
MODEL:
  ROI_BOX_HEAD:
    USE_ZEROSHOT_CLS: false
    FED_LOSS_FREQ_WEIGHT: 0.5
  WEIGHTS: "models/swin_large_patch4_window12_384_22k.pkl"
  BACKBONE:
    NAME: build_swintransformer_fpn_backbone
  SWIN:
    SIZE: L-22k-384
  FPN:
    IN_FEATURES: ["swin1", "swin2", "swin3"]
SOLVER:
  MAX_ITER: 180000
  CHECKPOINT_PERIOD: 30000
  IMS_PER_BATCH: 16
  BASE_LR: 0.0001
  MODEL_EMA: 0.999
INPUT:
  TRAIN_SIZE: 896
DATASETS:
  TRAIN: ("lvis_v1_train",)
  TEST: ("lvis_v1_val",)



AP | AP50 | AP75 | APs | APm | APl | APr | APc | APf
-- | -- | -- | -- | -- | -- | -- | -- | --
48.188 | 62.969 | 50.685 | 36.218 | 58.555 | 67.794 | 41.308 | 48.5 | 50.864
42.946 | 60.722 | 46.254 | 30.305 | 54.546 | 63.739 | 36.786 | 43.958 | 44.523

**Comment by yoctta** [comment]
Hi @leaf1170124460, I find the possible reason. The default config has been changed to "_C.INPUT.USE_COPY_METHOD = 'self_copy'". If you want to replicate the baseline, please set INPUT.USE_COPY_METHOD="none"

**Comment by Z-MU-Z** [comment]
> we use configs/Base-C2_L_R5021k_640b64_4x.yaml with 90k iters to train the base model. Which doesn't gain from more training steps. And use the model to initialize Xpaste training.
 @yoctta so what about copy paste? is that also use the  base model to initialize training?

**Comment by yoctta** [comment]
we didn't use self copy to train the base model

**Comment by Jiahao000** [comment]
Hi @yoctta, I met the same problem reproducing the baseline results in the paper. The results I reproduced using `configs/Base-C2_L_R5021k_640b64_4x.yaml` are similar to @DavideA:

|   Model  |   source   | box AP | mask AP | box APr | mask APr |
|:--------:|:----------:|:------:|:-------:|:-------:|:--------:|
| baseline | checkpoint |  34.5  |   30.8  |   24.0  |   21.6   |
| baseline | retraining |  33.7  |   30.0  |   20.8  |   18.5   |

I followed the instruction to use `configs/Base-C2_L_R5021k_640b64_4x.yaml` with 90k iters and a batch size of 64 to train the base model. However, the results seem worse than expected. What am I missing? Thanks in advance.

**Comment by yoctta** [comment]
MODEL_EMA: 0.999

**Comment by Jiahao000** [comment]
Thanks for your quick reply! Will have a try.

**Comment by saksham-s** [comment]
Hi @yoctta , I am also trying to run the baseline Centernet2 training using R50 backbone and while the training runs it keep printing "error in convert". Any clue how to fix this. I am using the configs/Base-C2_L_R5021k_640b64_4x.yaml

I was able to pinpoint its happening in this call 'gt_masks':inst.get('gt_masks').tensor.numpy() L23 in custom_copypaste.py. Did anyone else fix this issue? In my understanding the inst.get('gt_masks') is a Polygon and doing .tensor.numpy() on it causes the issue. I am also wondering why its using custom_copypaste.py in the first place.

Also if @DavideA @Jiahao000 @Z-MU-Z faced and corrected it I would appreciate any pointers.

**Comment by Z-MU-Z** [comment]
@saksham-s I encountered the same problem "error in convert". I obtained the above results without dealing with it.

**Comment by Jiahao000** [comment]
Hi @yoctta, I added  `MODEL_EMA: 0.999` in `configs/Base-C2_L_R5021k_640b64_4x.yaml` to run the baseline model. However, there is no significant increase in the results and they still seem lower than expected:

|   Model  |   source   | box AP | mask AP | box APr | mask APr |
|:--------:|:----------:|:------:|:-------:|:-------:|:--------:|
| baseline | checkpoint |  34.5  |   30.8  |   24.0  |   21.6   |
| baseline | retraining |  34.0  |   30.1  |   20.7  |   18.1   |

Could you help me find what is the problem to reproduce the baseline results?

**Comment by saksham-s** [comment]
Hi @yoctta I am facing a similar issue, when following each of the steps in the repo to setup the data and running the following command `python train_net.py --config-file configs/Xpaste_R50.yaml --dist-url tcp://0.0.0.0:12345 --num-gpus 8 --resume OUTPUT_DIR xpaste_approach SEED 42` I get 34.3379 as the Box AP and 30.1842 as the mask AP. Given its only using generated data I was expecting the number to be similar to row 1 of Table 2.

**Comment by yoctta** [comment]
Did you pretrain the CenterNet2 model without X-Paste? You can try to use the base model as pretrain

**Comment by Ab-34** [comment]
> Hi @yoctta , I am also trying to run the baseline Centernet2 training using R50 backbone and while the training runs it keep printing "error in convert". Any clue how to fix this. I am using the configs/Base-C2_L_R5021k_640b64_4x.yaml
> 
> I was able to pinpoint its happening in this call 'gt_masks':inst.get('gt_masks').tensor.numpy() L23 in custom_copypaste.py. Did anyone else fix this issue? In my understanding the inst.get('gt_masks') is a Polygon and doing .tensor.numpy() on it causes the issue. I am also wondering why its using custom_copypaste.py in the first place.
> 
> Also if @DavideA @Jiahao000 @Z-MU-Z faced and corrected it I would appreciate any pointers.

Could you please tell me how you fixed this? I am also running into this error, and then the process stops with the following message:
![image](https://github.com/yoctta/XPaste/assets/55902724/e93d6fbb-17fb-4c7b-a800-5c006521d760)
